### PR TITLE
Public code release

### PR DESCRIPTION
**Pull request by GabPrato**

### CODE REVIEW

File: accelerate_patch/apply_patch.sh
Comment (AndreasMadsen) [inline accelerate_patch/apply_patch.sh]: If using a patch is really the intended solution. Then you should pin accelerate to a specific commit.

File: pyproject.toml
Comment (AndreasMadsen) [inline pyproject.toml]: Not found.

File: pyproject.toml
Comment (AndreasMadsen) [inline pyproject.toml:8]: I don't think this setup is correct. You appear to use the `src-layout` (https://setuptools.pypa.io/en/latest/userguide/package_discovery.html#src-layout) but have not places your content in a named subdirectory. I would guess this code installs 3 root packages named `__init__`, `dataloader`, and `run`. When it properly should install just one package named `epik_eval`.

Also, consider using a `__main__.py` instead of a `run.py`.

**Review by AndreasMadsen — COMMENTED** [review]
**This is not a complete review. I can review again after the refactor.**

I don't think `dataset_generator`, `generate_paper_plots`, and `compute_paper_metrics`, are well suited as notebooks. They appear to be a collection of several independent scripts. I think the code would be better structured as such.

It's also not possible for me to properly review that code, with the GitHub interface as it's notebooks.

The README is missing instructions regarding the notebook code. Which appears to be essential to generate the paper results.

Regarding the accelerator patch. Have you tried upstreaming the changes? If not, you should try. If yes, maybe it's better to point to your repo-brach as a dependency. It would be less confusing.

File: accelerate_patch/apply_patch.sh
Comment (GabPrato) [reply accelerate_patch/apply_patch.sh]: The "patch" is not meant to fix an issue with accelerate's DataLoader, rather it adds support for the type of batch expected by our benchmark.

File: pyproject.toml
Comment (GabPrato) [reply pyproject.toml]: Fixed

File: pyproject.toml
Comment (GabPrato) [reply pyproject.toml:8]: Fixed

**Comment by GabPrato** [comment]
@AndreasMadsen addressed all mentioned issues. Code is ready for a second review.

File: setup.py
Comment (AndreasMadsen) [inline setup.py]: You should not have both a `setup.py` and `pyproject.toml` file. `pyproject.toml` have replaced `setup.py` in modern python.

File: setup.py
Comment (AndreasMadsen) [inline setup.py]: You can make directories before the file is written, not on install. Install side-effects are deprecated due to the security risk.

File: accelerate_patch/accelerator.patch
Comment (AndreasMadsen) [inline accelerate_patch/accelerator.patch]: Why is this required? Should the training batch_size not be less than the eval batch size?

File: accelerate_patch/accelerator.patch
Comment (AndreasMadsen) [inline accelerate_patch/accelerator.patch]: I think you can just set `batch_sizes[0] = batch_sizes[0] // deepspeed_plugin.deepspeed_config["gradient_accumulation_steps"]` when configuring accelerate. The values will be the same. This way you can avoid the patch.

File: accelerate_patch/data_loader.patch
Comment (AndreasMadsen) [inline accelerate_patch/data_loader.patch]: All this code is just transforming batch. It's unclear why it cannot be implemented with `slice_fn_for_dispatch `. If not, you should open an issue or contribute a PR.

File: configs/flan-t5-base/run_segmented.sh
Comment (AndreasMadsen) [inline configs/flan-t5-base/run_segmented.sh:1]: I don't know accelerate well. Why are some parameters in `.yaml` and others as command line arguments. Should they not all be in `.yaml`?

File: scripts/generate_dataset.py
Comment (AndreasMadsen) [inline scripts/generate_dataset.py]: use `getattr(args, f'{_set}_samples_per_task')`.

File: scripts/generate_dataset.py
Comment (AndreasMadsen) [inline scripts/generate_dataset.py]: Make an array and index it. `fns = [task1_sample, task2_sample, ...]` and `fns[task - 1]`.

File: scripts/generate_dataset.py
Comment (AndreasMadsen) [inline scripts/generate_dataset.py]: Same

File: scripts/generate_dataset.py
Comment (AndreasMadsen) [inline scripts/generate_dataset.py]: What do these `i` refer to? It looks like they are meant to refer to the task. However, that will only be the case if all tasks are generated. If this is the case, you should fix it such that the indices don't change depending on what tasks are enabled.

File: scripts/generate_paper_plots.py
Comment (AndreasMadsen) [inline scripts/generate_paper_plots.py]: Side-effect. Better to close the plot in the caller or the function that created the plot that is now being closed.

File: scripts/generate_paper_plots.py
Comment (AndreasMadsen) [inline scripts/generate_paper_plots.py:432]: Not a required change since they are just plot. But it's usually nice to be able to generate just one plot. As generating all is hard to debug. I suggest code for each plot is moved to it's own file.

File: scripts/README.md
Comment (AndreasMadsen) [inline scripts/README.md]: The dataset is core to the repo and the code. You should move the docs to the main README and document it more.

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py:33]: If `pad_token` are not already part of the model. Why would the token mask be handled correctly?

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py:184]: Why does the `train_dataloader` not produce mini-batches? If there is a good reason, you should document it.

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py]: If I understand correctly. This will give the same response every time. Calling 'iter()' is supposed to be a stateless operation. If it is stateless, why not cache it outside the while loop? If it is not stateless, this is an anti-pattern that should be fixed.

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py]: Why `any()` should it not be `all()` or check only the main process?
Could you not check if `next(iter)` throws `StopIteration`?

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py:254]: I think you only want to gather in the main process, right?

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py]: `pd.DataFrame.query` is pretty good for this.

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py]: Minor: replace `parents[0]` with `parent`.

File: src/epik_eval/__main__.py
Comment (AndreasMadsen) [inline src/epik_eval/__main__.py]: Minor: since you are already using path lib, you can use `path.with_stem(f"{path.stem}_epoch{epoch}")`.

File: src/epik_eval/dataloader.py
Comment (AndreasMadsen) [inline src/epik_eval/dataloader.py:119]: It's suspicious that these are called train when it's an eval dataloader.

File: src/epik_eval/dataloader.py
Comment (AndreasMadsen) [inline src/epik_eval/dataloader.py:191]: `return iter(self.dataset)` instead. Avoids keeping a generator scope and object.

File: src/epik_eval/dataloader.py
Comment (AndreasMadsen) [inline src/epik_eval/dataloader.py]: An `Iterable` does not finish, it is stateless. An `iterator` finish. I think there is a mix of responsibility here, which is properly what is causing you issues.

I guess, the iterator that this __iter__ returns, should either yield all observations from every batch or it should yield all batches.

**Review by AndreasMadsen — COMMENTED** [review]
In addition to documenting the code. I think it would be beneficial to document the dataset. How to decrypt it? What is the format? It likely that someone would not use the code but just the dataset.

You need to document the CLI for creating the dataset in details. Preferably also the metrics and plots (I don't think you need to document that in details).

I terms of the code, I'm very confused about the management of iterables. There seams to be some important dynamic with how batches are managed. I guess it's important that some observations are evaluated together, but it's not clear why. This seams to relate to the accelerator patch, the `while True` and the not-stateless iterable. It's all looks like an anti-pattern. Maybe there is some highly specific reason for all this, but it does not appear to be documented anywhere and I worry it's rather due to a common misunderstand regarding the difference between iterables and iterators.

File: accelerate_patch/accelerator.patch
Comment (GabPrato) [reply accelerate_patch/accelerator.patch]: Expecting the train batch size to always be smaller than the eval batch size isn't reliable. Although you could argue that expecting the train batch size to always be first in the list of batch sizes is also not reliable. Given that it is in my code though, this is more robust.

File: accelerate_patch/accelerator.patch
Comment (GabPrato) [reply accelerate_patch/accelerator.patch]: I prefer patching for the reason mentioned above. While at it, might as well patch this part.

File: accelerate_patch/data_loader.patch
Comment (GabPrato) [reply accelerate_patch/data_loader.patch]: I need to patch `data_loader.py` for the code above this. While at it, having this code here or in `__main__.py` makes no difference to me.

File: configs/flan-t5-base/run_segmented.sh
Comment (GabPrato) [reply configs/flan-t5-base/run_segmented.sh:1]: The `.yaml` contains the config for accelerate, while the command line arguments are for my `__main__.py` script.

File: scripts/generate_dataset.py
Comment (GabPrato) [reply scripts/generate_dataset.py]: Called `eval` once at the beginning of the main loop instead.

File: scripts/generate_dataset.py
Comment (GabPrato) [reply scripts/generate_dataset.py]: They are the doc's id. I renamed the variable.

File: src/epik_eval/__main__.py
Comment (GabPrato) [reply src/epik_eval/__main__.py:33]: The tokenizer does mask these tokens.

File: src/epik_eval/__main__.py
Comment (GabPrato) [reply src/epik_eval/__main__.py]: I've explained the design decisions all the way below this pull request. Let me know if you still have comments or questions.

File: src/epik_eval/__main__.py
Comment (GabPrato) [reply src/epik_eval/__main__.py:184]: The minibatch size is per device. If I were to initialize the dataloader with the minibatch size instead of the total batch size, it would divide the minibatch size by the number of device, which would change how I defined it and thus its value in each config.

File: src/epik_eval/__main__.py
Comment (GabPrato) [reply src/epik_eval/__main__.py:254]: `gather` needs to be called on all devices, or else it blocks.

File: src/epik_eval/dataloader.py
Comment (GabPrato) [reply src/epik_eval/dataloader.py:119]: During evaluation, I also measure how well the model memorized training samples.

File: src/epik_eval/dataloader.py
Comment (GabPrato) [reply src/epik_eval/dataloader.py:191]: This is placeholder code, as the dataloader gets replaced by Accelerate's.

File: src/epik_eval/dataloader.py
Comment (GabPrato) [reply src/epik_eval/dataloader.py]: See my comment below explaining the design.

**Comment by GabPrato** [comment]
@AndreasMadsen Addressed your latest review. Let me know if you have any further comments or questions.

> In addition to documenting the code. I think it would be beneficial to document the dataset. How to decrypt it? What is the format? It likely that someone would not use the code but just the dataset.

Done.
> You need to document the CLI for creating the dataset in details. Preferably also the metrics and plots (I don't think you need to document that in details).

Done.
> I terms of the code, I'm very confused about the management of iterables. There seams to be some important dynamic with how batches are managed. I guess it's important that some observations are evaluated together, but it's not clear why. This seams to relate to the accelerator patch, the `while True` and the not-stateless iterable. It's all looks like an anti-pattern. Maybe there is some highly specific reason for all this, but it does not appear to be documented anywhere and I worry it's rather due to a common misunderstand regarding the difference between iterables and iterators.

The issue is with the way the accelerate dataloader works. In `EvaluationDataset`, if I yield all of samples at once, Accelerate's dataloader will mix them up when yielding batches to my main script, even if you specify not to shuffle samples. Thus, samples in batches will have different lengths. In my training loop, I solve this by padding all samples up to the same length in `TrainDataset`, and then when I receive a batch in `__main__.py`, I cut up to the longest length, removing the extra padding. For evaluation though, since samples don't need to be shuffled, rather than cutting up to the longest length, it is much more efficient to prepare the batches only once in `EvaluationDataset`, pad to the correct length, and then yield samples only in that batch to Accelerate's Dataloader. This explains the need of a `while True` in the main loop, because now I have to call the __iter__ function once for every batch in my evaluation set.

**Comment by AndreasMadsen** [comment]
I'm supported to review the code based on if it's understandable and extendable. Right now, I would say "no" to both.

**Understandable:** The `while True: next(iter(d))` is an anti-pattern. This is not how we iterate on iterables. For example, if `d` was a list, you would just get the first element every time. That is the behavior I expect regardless of which data-structure supports `d`. However, because `d` mutates on `__iter__` we get different responses. You really need to not be mutating in  `__iter__`.

**Extendable:** The code requires patching a specific version of accelerate. That means that someone else can't apply this code to their project and use default accelerate or any other version of accelerate.

---

I would also recommend to never use `eval`, while I can see it properly doesn't present a security risk in this case. It just looks bad, as it's generally an anti-pattern, and half of the our review guidelines are about appearance. It's also slow and makes it harder to debug the code. This is such a minor issue compared to the other two, I don't want to focus on this. But I was surprised you didn't remove them all, so I thought it worth mentioning.

**Comment by GabPrato** [comment]
Switched the way the evaluation loop works. Now I pad the entire dataset to the same length, return the length and in the evaluation loop, I cut out the extra padding up to the longest prompt in the current batch.

Also replaced the only remaining `eval()` call.

With respect to the accelerate patch, their `data_loader.py` doesn't do what I want:
```python
if not self._drop_last and stop_iteration and observed_batch_size % self.state.num_processes != 0:
    # If the last batch is not complete, let's add the first batch to it.
    batch = concatenate([batch, first_batch], dim=0)
    # Batch size computation above is wrong, it's off by 1 so we fix it.
    batch_size += 1
```
During evaluation, I don't want to drop the last batch or add the first batch in the last batch. So the only way to avoid this would be to yield additional samples in my evaluation dataset, such that there is enough to "fill" the last batch (`len(eval_dataset) % batch_size == 0`). But then I would need to mark these extra samples as duplicates or samples to be discarded (as I am doing right now, but in the data loader) for the main loop to not take them into consideration. It feels weird to me for a dataset class to handle this, rather than the data loader. I understand that this forces users to stick with a specific accelerate version, but the way I see it, nobody is going to use this code. So I don't think it is worth spending time on this.

**Comment by AndreasMadsen** [comment]
According to https://github.com/huggingface/accelerate/issues/125

> For evaluation, you can just truncate the arrays of predictions to the length of your validation dataset, to ignore those added results.

**Comment by GabPrato** [comment]
Evaluation dataset now adds duplicates such that the evaluation dataset size is a multiple of the evaluation batch size. Duplicates are eliminated in the main loop as before.

DeepSpeed support has also been removed. All configs weren't using DeepSpeed anyways.

**Review by AndreasMadsen — APPROVED** [review]
I have approved the code. You will need a second review.

For public release this is fine. For anonymous submissions remember to delete the LICENSE and update `pyproject.toml`. 

I think the README could be more clear about how exactly to generate each plot presented in the paper.

**Review by jhuang265 — APPROVED** [review]
LGTM. Like andreas mentioned, maybe replace the figures with the code to generate the figures, but that's not really important.

**Review by jhuang265 — APPROVED** [review]

**Review by jhuang265 — APPROVED** [review]

**Pull request merged**
